cmd/clusterctl/cmd/config_repositories_test.go

```diff
@@ -0,0 +1,108 @@
+/*
+Copyright 2020 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package cmd
+
+import (
+	"bytes"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"testing"
+
+	. "github.com/onsi/gomega"
+)
+
+func Test_runGetRepositories(t *testing.T) {
+	t.Run("prints output", func(t *testing.T) {
+		g := NewWithT(t)
+
+		tmpDir, err := ioutil.TempDir("", "cc")
+		g.Expect(err).NotTo(HaveOccurred())
+		defer os.RemoveAll(tmpDir)
+
+		path := filepath.Join(tmpDir, "clusterctl.yaml")
+		g.Expect(ioutil.WriteFile(path, []byte(template), 0644)).To(Succeed())
+
+		buf := bytes.NewBufferString("")
+		g.Expect(runGetRepositories(path, buf)).To(Succeed())
+
+		out, err := ioutil.ReadAll(buf)
+		g.Expect(err).ToNot(HaveOccurred())
+		g.Expect(string(out)).To(Equal(expectedOutput))
+	})
+
+	t.Run("returns error for bad cfgFile path", func(t *testing.T) {
+		//TODO: Remove skip when
+		//https://github.com/kubernetes-sigs/cluster-api/issues/2897 is fixed
+		t.Skip()
+		g := NewWithT(t)
+		buf := bytes.NewBufferString("")
+		g.Expect(runGetRepositories("do-not-exist", buf)).ToNot(Succeed())
+	})
+
+	t.Run("returns error for nil writer", func(t *testing.T) {
+		g := NewWithT(t)
+		g.Expect(runGetRepositories("do-exist", nil)).ToNot(Succeed())
+	})
+
+	t.Run("returns error for bad template", func(t *testing.T) {
+		g := NewWithT(t)
+
+		tmpDir, err := ioutil.TempDir("", "cc")
+		g.Expect(err).NotTo(HaveOccurred())
+		defer os.RemoveAll(tmpDir)
+
+		path := filepath.Join(tmpDir, "clusterctl.yaml")
+		g.Expect(ioutil.WriteFile(path, []byte("providers: foobar"), 0644)).To(Succeed())
+
+		buf := bytes.NewBufferString("")
+		g.Expect(runGetRepositories(path, buf)).ToNot(Succeed())
+	})
+}
+
+var template = `---
+providers:
+  # add a custom provider
+  - name: "my-infra-provider"
+    url: "/home/.cluster-api/overrides/infrastructure-docker/latest/infrastructure-components.yaml"
+    type: "InfrastructureProvider"
+  # add a custom provider
+  - name: "another-provider"
+    url: "./bootstrap-components.yaml"
+    type: "BootstrapProvider"
+  # bad url
+  - name: "aws"
+    url: "my-aws-infrastructure-components.yaml"
+    type: "InfrastructureProvider"
+  # override a pre-defined provider
```

Comment:
This feels more like testing the client `GetProvidersConfig` than this printer but can't think of an alternative given the client is also created inside this function.

Reply:
Yeah... I added this test because it is testing behavior affecting the user experience. I was thinking of other ways to test the output but at the same time didn't want to make sweeping changes to this cmd unless the community thinks that's ok to do 😉 

On the bright side I found issue #2897 when testing the real client 😄

Reply:
Nice!